lib/streamlit/delta_generator.py

```diff
@@ -777,11 +781,21 @@ def _value_or_dg(value, dg):
     return value
 
 
-def _enqueue_message(msg):
-    """Enqueues a ForwardMsg proto to send to the app."""
+def ensure_report_context(default_return=None):
     ctx = get_report_ctx()
 
     if ctx is None:
-        raise NoSessionContext()
+        if default_return is None:
+            raise NoSessionContext()
+        else:
+            # For scripts not running through streamlit
```

Comment:
It looks like the reason we're still tracking this is for when the user is running the app bare (aka `python app.py` instead of `streamlit run app.py`); we're returning what is sometimes a DeltaGenerator from this method, because both DGs and ReportContexts now have a dg_stack that can be referenced?

This pattern seems quite strange to me, that "ensure_report_context" sometimes returns a DG. I'm actually thinking that in the case of the script being run bare, the "with" stack doesn't matter at all? Because all st function calls are supposed to be no-ops anyways (widgets do need to return the correct value, but that should have no bearing on which DG is being used)

If we DO want to rely on DG.dg_stack for a bare run, note that it's currently doing a "wrong" thing of always referring to itself. That is, if this is run bare:
```
with DG1:
    st.write('blah')
```
At the point of st.write(), we'll look up `_active_dg` and find it to be MAIN, not DG1, I think. Again, it doesn't cause any errors now because everything  a no-op anyways; but we should be explicit that DG.dg_stack is not a reliable source of which "with" level we're working at.

Reply:
Anyways, my suggestion for this PR would be to make this more explicit; instead of having the ensure_report_context function and having a fallback dg_stack, we could just no-op appropriately when get_report_ctx() is none. E.g.:

```
    def __enter__(self):
        # with block started
        ctx = get_report_ctx()
        if ctx:
            ctx.dg_stack.append(self)
```

Reply:
Good to know, yea, I was just grasping at straws to get the tests to pass a little bit. I'll make the update.